File: combinations.cpp
//Runtime: 73 ms
class Solution {
public:

    void solve(vector<vector<int> > &res, vector<int>&temp, int l, int c, int k,int n)
    {
        if(c == k)
        {
            res.push_back(temp);
            return;
        }

        for(int i=l;i<=n;i++)
        {
            temp.push_back(i);
            solve(res, temp, i+1,c+1, k, n);
            temp.pop_back();
        }
    }

    vector<vector<int>> combine(int n, int k) {
        vector<vector<int> > res;
        vector<int>temp;
        solve(res, temp, 1, 0, k, n);
        return res;
    }
};

File: combinations.py
# Runtime: 262 ms
class Solution(object):

    def solve(self, res, t, c, k, n):
        if len(t) == k:
            res.append(t)
            return
        if k > len(t) + n - c:
            return
        for i in xrange(c, n):
            self.solve(res, t + [i], i + 1, k, n)

    def combine(self, n, k):
        res = []
        t = []

        self.solve(res, t, 1, k, n + 1)
        return res
